File: lab06/AVL_tree.c
#include<stdio.h>
#include<stdlib.h>
#define max(a,b)  (((a) > (b)) ? (a) : (b))
#define min(a,b)  (((a) < (b)) ? (a) : (b))

FILE *fin;
FILE *fout;

struct AVLNode;
typedef struct AVLNode *Position;
typedef struct AVLNode *AVLTree;
typedef int ElementType;

typedef struct AVLNode{
	ElementType element;
	AVLTree left, right;
	int height;
}AVLNode;

Position SingleRotateWithLeft(Position node);
Position SingleRotateWithRight(Position node);
Position DoubleRotateWithLeft(Position node);
Position DoubleRotateWithRight(Position node);

Position selectRotation(Position node);
int setHeight(Position node);

// AVL tree에 X값을 넣는 함수
AVLTree Insert(ElementType X, AVLTree T){
	// 넣을 위치에 올 때 
	if(T == NULL){
		T = (AVLTree)malloc(sizeof(AVLNode));
		T->element = X;
		T->left = NULL;
		T->right = NULL;
		T->height = 0;
	}
	// 현재 node보다 작을 때 
	else if(T->element > X){
		T->left = Insert(X, T->left);
		T->height = setHeight(T);
		T = selectRotation(T);	
	}
	// 현재 node보다 클 때
	else if(T->element < X){
		T->right = Insert(X, T->right);
		T->height = setHeight(T);
		T = selectRotation(T);
	}
	// 이미 X가 존재할 때
	else{
		fprintf(fout, "insertion error : %d is already in the tree!\n", X);
	}
	return T;
};

// AVL tree에서 X값을 제거하는 함수
AVLTree Delete(ElementType X, AVLTree T){
	// tree에 X가 없을 떄
	if(T==NULL){
		fprintf(fout, "deletion error : %d is not in the tree!\n", X);
		return T;
	}
	// 현재 node보다 작을 때 
	else if(T->element > X) T->left = Delete(X, T->left);
	// 현재 node보다 클 때
	else if(T->element < X) T->right = Delete(X, T->right);
	
	// 찾았을 때
	else{
		Position temp;
		// child node가 오른쪽에만 있거나 둘 다 없는 경우
		if(T->left ==NULL){
			temp = T->right;
			free(T);
			return temp;
		}
		// child node가 왼쪽에만 있는 경우
		else if(T->right ==NULL){
			temp = T->left;
			free(T);
			return temp;
		}
		// 둘 다 있는 경우,
		else{
			temp = T->right;
			while(temp->left != NULL) temp = temp->left;
			T->element = temp->element;
			T->right = Delete(temp->element, T->right);
		}
	} 
	// Tree 높이 설정 및 회전
	T->height = setHeight(T);
	T = selectRotation(T);
	return T;
	};

// Case : LL
// node를 root로 뒀을 때 root -> root right로 / root left -> root로 / root left right -> root right로 이동
Position SingleRotateWithLeft(Position node){
	Position temp = node -> left;
	node->left = temp->right;
	temp->right = node;
	node->height = setHeight(node);
  temp->height = setHeight(temp);
	return temp;
};

// Case : RR
// node를 root로 뒀을 때 root -> root left로 / root right -> root로 / root right left -> root right로 이동
Position SingleRotateWithRight(Position node){
	Position temp = node -> right;
	node->right = temp->left;
	temp->left = node;
	node->height = setHeight(node);
  temp->height = setHeight(temp);
	return temp;
};

// Case : LR
// node를 root로 뒀을 때 root left에 대해 SingleRotateWithRight 진행, 이후 root에서 SingleRotateWithLeft
Position DoubleRotateWithLeft(Position node){
	node -> left = SingleRotateWithRight(node->left);
	return SingleRotateWithLeft(node);
};

// Case : RL
// node를 root로 뒀을 때 root right에 대해 SingleRotateWithLeft 진행, 이후 root에서 SingleRotateWithRight
Position DoubleRotateWithRight(Position node){
	node->right =SingleRotateWithLeft(node->right);
	return SingleRotateWithRight(node);
};

// insert, delete, rotate 이후 달라진 높이를 재설정해주는 함수
int setHeight(Position node) {
	// child가 2개일 때, 더 깊은 쪽 선택
	if(node->left != NULL && node->right != NULL){
		return max(node->left->height, node->right->height) + 1;
	}
	// left만 있을 때
	else if(node->left != NULL){
		return node->left->height + 1;
	}
	// right만 있을 때
	else if(node->right != NULL){
		return node->right->height + 1;
	}
	// 둘 다 없으면 leaf node
	else{
		return 0;
	}
};

// 어떤 Rotation 함수를 실행할지 선택해주는 함수
Position selectRotation(Position node){
	// child가 없을 때
	if(node->left == NULL && node->right == NULL) return node;
	// right child가 없을 때, left depth가 2 이상이면,
	else if(node->right == NULL){
		if(node->height >= 2) node = SingleRotateWithLeft(node);
	}
	// left child가 없을 때, right depth가 2 이상이면,
	else if(node->left == NULL){
		if(node->height >= 2) node = SingleRotateWithRight(node);
	}
	// 양쪽 child가 다 있을 때 
	else{
		// depth 차이가 2 이상이면 
		if(max(node->left->height, node->right->height)- min(node->left->height, node->right->height) >= 2){
			Position temp2;
			// // left가 더 깊을 때
			if(node->left->height > node->right->height){
				temp2 = node->left;
				// rotation case 판단
				// left subtree에서 right가 left보다 2이상 깊으면 double
				if((temp2->left == NULL && temp2->right->height == 1) || 
				(temp2->left != NULL && temp2->right->height - temp2->left->height == 2)) node = DoubleRotateWithLeft(node);
				// 아닐 경우 single
				else node = SingleRotateWithLeft(node);
			}
			// // right가 더 깊을 때 
			else{
				temp2 = node->right;
				// rotation case 판단
				if((temp2->right == NULL && temp2->left->height == 1) || 
				(temp2->right != NULL && temp2->left->height - temp2->right->height == 2)) node = DoubleRotateWithRight(node);
				else node = SingleRotateWithRight(node);
			}
		}
	}
	return node;
};

// Inorder순으로 출력해주는 함수
void PrintInorder(AVLTree T){
	if(T == NULL) return;
	PrintInorder(T -> left);
	fprintf(fout, "%d(%d) ", T -> element, T->height);
  PrintInorder(T -> right);
};

// Tree 전체를 제거해주는 함수
void DeleteTree(AVLTree T){
	if(T == NULL) return;
	DeleteTree(T -> left);
	DeleteTree(T -> right);
	free(T);
};

int main(int argc, char *argv[]){
	fin = fopen(argv[1], "r");
	fout = fopen(argv[2], "w");
	AVLTree Tree = NULL;
	char cv;
	int key;

	while(!feof(fin)){
		fscanf(fin, "%c", &cv);
		switch(cv){
			case 'i':
				fscanf(fin, "%d\n", &key);
				Tree = Insert(key, Tree);
				break;
			case 'd':
				fscanf(fin, "%d\n", &key);
				Tree = Delete(key, Tree);
				break;
		}
		PrintInorder(Tree);
		fprintf(fout, "\n");
	}

	DeleteTree(Tree);
	fclose(fin);
	fclose(fout);

	return 0;
}

File: assignment1/AVLTree.py
import sys

# node class
class AVLNode:
  def __init__(self, key, value):
    self.key = key
    self.value = value
    self.left = None
    self.right = None
    self.height = 0

# single rotate(Left)
def single_rotate_with_left(node):
  temp = node.right
  node.right = temp.left
  temp.left = node
  node.height = set_height(node)
  temp.height = set_height(temp)
  return temp

# single rotate(Right)
def single_rotate_with_right(node):
  temp = node.left
  node.left = temp.right
  temp.right = node
  node.height = set_height(node)
  temp.height = set_height(temp)
  return temp

# double rotate(Left-Rgiht)
def double_rotate_with_left(node):
  node.right = single_rotate_with_right(node.right)
  return single_rotate_with_left(node)

# double rotate(Right-Left)
def double_rotate_with_right(node):  
  node.left = single_rotate_with_left(node.left)
  return single_rotate_with_right(node)

def get_balance_factor(node):
  if node.left and node.right :
    return node.left.height - node.right.height
  elif node.left:
    return node.left.height + 1
  elif node.right:
    return -(node.right.height + 1)
  else:
    return 0

def select_rotation(node):
  balance_factor = get_balance_factor(node)
  # left
  if balance_factor < -1:
    balance_factor = get_balance_factor(node.right)
    if balance_factor > 0:
      node = double_rotate_with_left(node)
    else:
      node = single_rotate_with_left(node)
      
  # right
  elif balance_factor > 1:
    balance_factor = get_balance_factor(node.left)
    if balance_factor < 0:
      node = double_rotate_with_right(node)
    else:
      node = single_rotate_with_right(node)
  return node

def set_height(node):
  if node.left and node.right:
    return max(node.left.height, node.right.height) + 1
  elif node.left:
    return node.left.height + 1
  elif node.right:
    return node.right.height + 1
  else:
    return 0

def insert(key, value, t):
  if t is None:
    t = AVLNode(key, value)
  elif t.key > key:
    t.left = insert(key, value , t.left)
  elif t.key < key:
    t.right = insert(key, value, t.right)
    
  t.height = set_height(t)
  t = select_rotation(t)
  return t

def delete(key, t):
  if t == None:
    return t
  elif t.key > key:
    t.left = delete(key, t.left)
  elif t.key < key:
    t.right = delete(key, t.right)
  else:
    temp = None
    if t.left is None:
      temp = t.right
      return temp
    elif t.right is None:
      temp = t.left
      return temp
    else:
      temp = t.right
      while temp.left is not None:
          temp = temp.left
      t.key = temp.key
      t.value = temp.value
      t.right = delete(temp.key, t.right)
  
  t.height = set_height(t)
  t = select_rotation(t)
  return t

def find(key, t):
  if t == None:
    return None
  elif t.key == key:
    return t
  elif t.key > key:
    return find(key, t.left)
  else:
    return find(key, t.right)

# def print_inorder(t, output2):
#   if t is None:
#     return
#   print_inorder(t.left, output2)
#   output2.write(str(t.key)+ "\t" + str(t.value)+ "\n")
#   print_inorder(t.right, output2)

def search(t, fin, fout):
  for line in fin:
    key = int(line.split()[0])
    node = find(key, t)
    if node:
      fout.write(str(node.key)+ "\t" + str(node.value)+ "\n")
    else:
      fout.write(str(key)+ "\tN/A\n")
    
def compare(fin, fout):
  for line1, line2 in zip(fin, fout):
    if line1.split()[0] != line2.split()[0] or line1.split()[1] != line2.split()[1]:
      print("Comparison failed.")
      return
    print("Comparison succeeded.")
    return
  
# main
# python ***.py
tree = None
while(True):
  command = int(input("1.Insertion, 2.Deletion, 3.Quit\n"))
  
  # insertion
  if(command == 1):
    file1 = input("insert file name\n")
    insert_result = input("output file name\n")
    with open(file1, 'r') as fin:
      for line in fin:
        key = int(line.split()[0])
        value = int(line.split()[1])
        tree = insert(key, value, tree)
    with open(file1, 'r') as fin, open(insert_result, 'w') as fout:
      search(tree, fin, fout)
    with open(file1, 'r') as fin, open(insert_result, 'r') as fout:
      compare(fin, fout)
  
  # deletion
  elif(command == 2):
    file2 = input("delete file name\n")
    delete_result = input("output file name\n")
    with open(file2, 'r') as fin:
      for line in fin:
        key = int(line.split()[0])
        tree = delete(key, tree)
    with open(file1, 'r') as fin, open(delete_result, 'w') as fout:
      search(tree, fin, fout)
    compare_filename = input("compare file name\n")
    with open(compare_filename, 'r') as fin, open(delete_result, 'r') as fout:
      compare(fin, fout)
  
  # quit
  elif(command == 3):
    sys.exit()